<a href="https://colab.research.google.com/github/rahul0xcdf/Python-Pattern-problems/blob/main/Tensorflow_first_run_DL.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [51]:
import pandas as pd
from sklearn.model_selection import train_test_split



In [52]:
df = pd.read_csv("/content/TF_DL/Churn.csv")


In [53]:
X = pd.get_dummies(df.drop(['Churn', 'Customer ID'], axis=1))
y = df['Churn'].apply(lambda x: 1 if x=='Yes' else 0)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.2)

In [7]:
X_train.head()

Senior Citizen  tenure  Monthly Charges  Gender_Female  Gender_Male  \
489                0       0            52.55           True        False   
4876               1      48           117.45           True        False   
2316               0      35            73.45          False         True   
2755               0       3            41.35          False         True   
4474               1      26            98.10           True        False   

      Partner_No  Partner_Yes  Dependents_No  Dependents_Yes  \
489        False         True          False            True   
4876       False         True           True           False   
2316       False         True          False            True   
2755        True        False           True           False   
4474       False         True           True           False   

      Phone Service_No  ...  Total Charges_995.35  Total Charges_996.45  \
489               True  ...                 False                 False   
4876             False  ...                 False                 False   
2316             False  ...                 False                 False   
2755              True  ...                 False                 False   
4474             False  ...                 False                 False   

      Total Charges_996.85  Total Charges_996.95  Total Charges_997.65  \
489                  False                 False                 False   
4876                 False                 False                 False   
2316                 False                 False                 False   
2755                 False                 False                 False   
4474                 False                 False                 False   

      Total Charges_997.75  Total Charges_998.1  Total Charges_999.45  \
489                  False                False                 False   
4876                 False                False                 False   
2316                 False                False                 False   
2755                 False                False                 False   
4474                 False                False                 False   

      Total Charges_999.8  Total Charges_999.9  
489                 False                False  
4876                False                False  
2316                False                False  
2755                False                False  
4474                False                False  

[5 rows x 6575 columns]

In [8]:
y_train.head()

489     0
4876    1
2316    0
2755    0
4474    0
Name: Churn, dtype: int64

## **Importing the dependencies**

In [10]:
from tensorflow.keras.models import Sequential, load_model
from tensorflow.keras.layers import Dense
from sklearn.metrics import accuracy_score

# Building the model a neural network with input layer, hidden layer, and the output layer.

In [30]:
model = Sequential()
model.add(Dense(units=8, activation='relu', input_dim=len(X_train.columns)))
model.add(Dense(units=16, activation='relu'))
model.add(Dense(units=1, activation='sigmoid'))

In [31]:
model.compile(loss='binary_crossentropy', optimizer='adam', metrics='accuracy')

In [32]:
import numpy as np
# Assuming X_train and y_train are NumPy arrays
X_train = X_train.astype(np.float32)
y_train = y_train.astype(np.float32)

model.fit(X_train, y_train, epochs=200, batch_size=32)

Epoch 1/200
177/177 [==============================] - 2s 5ms/step - loss: 0.5052 - accuracy: 0.7446
Epoch 2/200
177/177 [==============================] - 1s 6ms/step - loss: 0.4377 - accuracy: 0.7988
Epoch 3/200
177/177 [==============================] - 1s 6ms/step - loss: 0.4217 - accuracy: 0.8021
Epoch 4/200
177/177 [==============================] - 1s 6ms/step - loss: 0.4140 - accuracy: 0.8055
Epoch 5/200
177/177 [==============================] - 1s 5ms/step - loss: 0.4024 - accuracy: 0.8117
Epoch 6/200
177/177 [==============================] - 1s 5ms/step - loss: 0.3962 - accuracy: 0.8158
Epoch 7/200
177/177 [==============================] - 1s 5ms/step - loss: 0.3773 - accuracy: 0.8257
Epoch 8/200
177/177 [==============================] - 1s 3ms/step - loss: 0.3630 - accuracy: 0.8394
Epoch 9/200
177/177 [==============================] - 1s 3ms/step - loss: 0.3478 - accuracy: 0.8470
Epoch 10/200
177/177 [==============================] - 1s 3ms/step - loss: 0.3220 - accura

In [33]:
import numpy as np

# Assuming X_test is a Pandas DataFrame
X_test = X_test.astype(np.float32)  # Convert all columns in X_test to float32

y_hat = model.predict(X_test)
y_hat = [0 if val < 0.5 else 1 for val in y_hat]

45/45 [==============================] - 0s 3ms/step


In [36]:
accuracy_score(y_test, y_hat)

0.6728176011355571

32 and 64 neurons - 0.8119233498935415

64 and 128 - 0.7381121362668559

16 and 32 (with adam) - 0.6728176011355571

# **Increasing the model capacity**

In [41]:
model = Sequential()
model.add(Dense(units=16, activation='relu', input_dim=len(X_train.columns)))
model.add(Dense(units=32, activation='relu'))
model.add(Dense(units=64, activation='relu'))
model.add(Dense(units=1, activation='sigmoid'))


In [48]:
import tensorflow as tf

# ... your existing code ...

optimizer = tf.keras.optimizers.SGD(learning_rate=0.001)
model.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=['accuracy'])

In [49]:
import numpy as np
# Assuming X_train and y_train are NumPy arrays
X_train = X_train.astype(np.float32)
y_train = y_train.astype(np.float32)

model.fit(X_train, y_train, epochs=200, batch_size=32)

Epoch 1/200
177/177 [==============================] - 1s 3ms/step - loss: 1.5057 - accuracy: 0.8687
Epoch 2/200
177/177 [==============================] - 1s 3ms/step - loss: 0.4245 - accuracy: 0.8254
Epoch 3/200
177/177 [==============================] - 1s 3ms/step - loss: 0.3852 - accuracy: 0.8413
Epoch 4/200
177/177 [==============================] - 1s 3ms/step - loss: 0.4013 - accuracy: 0.8332
Epoch 5/200
177/177 [==============================] - 1s 3ms/step - loss: 0.3896 - accuracy: 0.8373
Epoch 6/200
177/177 [==============================] - 1s 3ms/step - loss: 0.3851 - accuracy: 0.8405
Epoch 7/200
177/177 [==============================] - 1s 3ms/step - loss: 0.3680 - accuracy: 0.8470
Epoch 8/200
177/177 [==============================] - 1s 3ms/step - loss: 0.3898 - accuracy: 0.8232
Epoch 9/200
177/177 [==============================] - 1s 3ms/step - loss: 0.3692 - accuracy: 0.8500
Epoch 10/200
177/177 [==============================] - 1s 3ms/step - loss: 0.3763 - accura

In [56]:
import numpy as np
from sklearn.metrics import accuracy_score  # Import the accuracy_score function

# Assuming X_test is a Pandas DataFrame
# X_test = X_test.astype(np.float32)  # Convert all columns in X_test to float32

y_hat = model.predict(X_test)
y_hat = [0 if val < 0.5 else 1 for val in y_hat]

# Use the imported accuracy_score function
accuracy = accuracy_score(y_test, y_hat)
print('Accuracy score: ', accuracy)

45/45 [==============================] - 0s 2ms/step
Accuracy score:  0.850958126330731
